### PR TITLE
Add Liquidator TG notifications for liquidations and swap

### PR DESCRIPTION
**Pull request by vitaliybezz**
Summary

- Add Telegram notifications for liquidation events (success, failure, swap issues)
- Add NonceTracker to prevent nonce collisions between oracle updates and liquidation transactions
- Add short human-readable asset names for notifications (e.g. BTC instead of nep141:btc.omft.near)

- - -
This change is [Reviewable](https://reviewable.io/reviews/Templar-Protocol/contracts/391)

### CODE REVIEW

File: service/liquidator/src/rpc.rs [hunk 1]
```diff
@@ -87,6 +91,36 @@ pub type BorrowPositions = HashMap<AccountId, BorrowPosition>;
...
+
+    /// Given an RPC-reported access key nonce, return the next safe nonce to use.
+    ///
+    /// Takes the max of the RPC nonce and our tracked nonce, then adds 1.
+    pub fn next_nonce(&self, rpc_access_key_nonce: u64) -> u64 {
+        let tracked = self.0.load(std::sync::atomic::Ordering::SeqCst);
+        let base = rpc_access_key_nonce.max(tracked);
+        let next = base + 1;
+        self.0.fetch_max(next, std::sync::atomic::Ordering::SeqCst);
+        next
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/rpc.rs]: Fixed. Replaced `load + fetch_max` with a `compare_exchange` CAS loop so each concurrent caller reserves a unique nonce. Added concurrent uniqueness test (100 threads).

File: service/liquidator/src/liquidator.rs [hunk 2]
```diff
@@ -194,14 +197,17 @@ impl Liquidator {
...
     ) -> Self {
+        let nonce_tracker = crate::rpc::NonceTracker::default();
         let scanner = scanner::MarketScanner::new(client.clone(), market.clone());
         let oracle_fetcher = oracle::OracleFetcher::new(
             client.clone(),
             hermes_url,
             redstone_gateway_url,
             proxy_oracle_cache,
             signer_for_oracle,
+            nonce_tracker.clone(),
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs]: Fixed. NonceTracker is now created once in `LiquidatorService::new()` and shared (via clone) to all Liquidators, swap providers, and oracle fetchers.

File: service/liquidator/src/config.rs [hunk 3]
```diff
@@ -273,6 +290,20 @@ impl Args {
...
+            tracing::info!("Telegram notifications disabled (no bot token)");
+            None
+        } else {
+            tracing::info!("Telegram notifications enabled");
+            Some(TelegramConfig {
+                bot_token: self.telegram_bot_token.clone(),
+                chat_id: self.telegram_chat_id.clone(),
+                thread_id: self.telegram_thread_id,
+            })
+        };
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/config.rs]: Fixed. Telegram is now only enabled when both `bot_token` and `chat_id` are non-empty (after trimming). Panics if only one is set. Added tests for all edge cases.

File: service/liquidator/src/swap/ref.rs [hunk 4]
```diff
@@ -473,7 +473,7 @@ impl SwapProvider for RefSwap {
                 .await?;
         }
 
-        let (nonce, block_hash) = get_access_key_data(&self.client, &self.signer).await?;
+        let (nonce, block_hash) = get_access_key_data(&self.client, &self.signer, None).await?;
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/swap/ref.rs]: Fixed. `RefSwap` now accepts a `NonceTracker` in its constructor and passes `Some(&self.nonce_tracker)` at both `get_access_key_data` call sites.

File: service/liquidator/src/swap/oneclick.rs [hunk 5]
```diff
@@ -602,7 +602,7 @@ impl OneClickSwap {
             "Storage deposit minimum from contract"
         );
 
-        let (nonce, block_hash) = get_access_key_data(&self.client, &self.signer).await?;
+        let (nonce, block_hash) = get_access_key_data(&self.client, &self.signer, None).await?;
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/swap/oneclick.rs]: Fixed. `OneClickSwap` now accepts a `NonceTracker` in its constructor and passes `Some(&self.nonce_tracker)` at all three `get_access_key_data` call sites.

File: service/liquidator/src/notifier.rs [hunk 6]
```diff
@@ -0,0 +1,225 @@
+//! Notification system for the liquidator bot.
+//!
+//! Sends alerts to Telegram when significant events occur:
+//! - Successful liquidations
+//! - Failed or skipped swaps (unsupported assets, errors)
+//!
+//! All methods are fire-and-forget: notification failures are logged
+//! but never block liquidation operations.
+
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs]: Fixed. All `notify_*` methods now use `tokio::spawn` internally via `spawn_send()` and return immediately. Callers cannot accidentally block the liquidation loop.

File: service/liquidator/src/format.rs [hunk 7]
```diff
@@ -4,6 +4,163 @@
...
+        ),
+        (
+            "nep141:eth-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599.omft.near",
+            "WBTC",
+        ),
+        (
+            "nep141:sol-5ce3bf3a31af18be40ba30f721101b4341690186.omft.near",
+            "USDC",
+        ),
+        // Native NEAR tokens
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/format.rs:36]: Fixed. Added tests for `short_asset_name` (known assets, nep245 wrapper stripping, Stellar suffix, generic nep245, nep141 fallback, truncation), `format_amount_short`, and `format_profit_short`.

File: service/liquidator/src/rpc.rs [hunk 8]
```diff
@@ -87,6 +91,36 @@ pub type BorrowPositions = HashMap<AccountId, BorrowPosition>;
...
 pub const DEFAULT_GAS: u64 = Gas::from_tgas(300).as_gas();
 
+/// Shared nonce tracker to prevent nonce collisions between concurrent transactions
+/// (e.g., oracle price updates and liquidation transactions).
+///
+/// Tracks the last nonce used by any transaction. When fetching a new nonce,
+/// returns `max(rpc_nonce, tracked_nonce) + 1` to avoid stale-nonce races
+/// where an RPC node hasn't indexed the latest transaction yet.
+#[derive(Debug, Clone, Default)]
+pub struct NonceTracker(Arc<AtomicU64>);
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/rpc.rs:101]: Fixed. Added 4 tests: basic increment, RPC jump forward, `record_used`, and concurrent uniqueness (100 threads spawned to verify no duplicate nonces).

File: service/liquidator/src/config.rs [hunk 9]
```diff
@@ -150,6 +155,18 @@ pub struct Args {
...
     #[arg(long, env = "SWAP_RETRY_BASE_DELAY_MS", default_value_t = 2000)]
     pub swap_retry_base_delay_ms: u64,
+
+    /// Telegram bot token for notifications (leave empty to disable)
+    #[arg(long, env = "TELEGRAM_BOT_TOKEN", default_value = "")]
+    pub telegram_bot_token: String,
+
+    /// Telegram chat/channel ID for notifications
+    #[arg(long, env = "TELEGRAM_CHAT_ID", default_value = "")]
+    pub telegram_chat_id: String,
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/config.rs:178]: Fixed. Both `bot_token` and `chat_id` are trimmed and must both be non-empty to enable Telegram. Panics with a clear message if only one is set. Added 6 tests covering all combinations.

File: service/liquidator/src/format.rs [hunk 10]
```diff
@@ -4,6 +4,163 @@
...
+        // nep141:something.omft.near already handled above
+        return rest.to_string();
+    }
+
+    // Last resort: truncate
+    if inner.len() > 20 {
+        format!("{}…", &inner[..17])
+    } else {
+        inner.to_string()
+    }
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/format.rs:128]: Fixed. Added a generic `nep245:{contract}:{token_id}` branch before the truncation fallback. It extracts `token_id` and recurses if it starts with `nep141:`/`nep245:`, otherwise uppercases it. e.g. `nep245:v2_1.omni.hot.tg:xlm` → `XLM`. Added regression test.

File: service/liquidator/src/liquidator.rs [hunk 11]
```diff
@@ -664,6 +675,59 @@ impl Liquidator {
...
+                        &format::format_profit_short(
+                            signed_profit,
+                            liquidation_amount.0,
+                            borrow_dec,
+                            &borrow_asset,
+                        ),
+                        None,
+                        dry_run,
+                    )
+                    .await;
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs]: Fixed. Notification profit now uses `liquidation_amount + gas_cost` as total cost, matching the profitability gate logic.

File: service/liquidator/src/notifier.rs [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        let tx_line = tx_hash.map_or(String::new(), |h| format!("\nTx: <code>{h}</code>"));
+
+        let message = format!(
+            "{prefix}✅ <b>Liquidation Executed</b>\n\
+             \n\
+             Market: <code>{market}</code>\n\
+             Borrower: <code>{borrower}</code>\n\
+             Sent: {send_amount}\n\
+             Received: {receive_amount}\n\
+             Profit: {profit}{tx_line}"
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs]: Fixed. All dynamic fields are now passed through `html_escape()` which replaces `<`, `>`, `&` with HTML entities. Surrounding `<b>`/`<code>` tags remain intact. Added test for escaping.

File: service/liquidator/src/notifier.rs [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        );
+
+        self.send(&message).await;
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs]: Fixed. `notify_*` methods now spawn the HTTP request via `tokio::spawn` and return immediately. The `send()` method runs on a background task.

File: service/liquidator/src/notifier.rs [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+    /// Failures are logged and swallowed — never propagated.
+    async fn send(&self, text: &str) {
+        let Some(config) = &self.telegram else {
+            return;
+        };
+
+        let url = format!(
+            "https://api.telegram.org/bot{}/sendMessage",
+            config.bot_token
+        );
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs:270]: Fixed. Now uses `e.without_url()` before logging the reqwest error, so the bot token in the URL is never written to logs.

File: service/liquidator/src/rpc.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/rpc.rs]: Fixed. Uses `compare_exchange` CAS loop for atomic nonce reservation. Verified with a 100-thread concurrency test — all nonces unique.

File: service/liquidator/src/swap/oneclick.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/swap/oneclick.rs]: Fixed. `OneClickSwap` now stores a `NonceTracker` field and uses it at all 3 `get_access_key_data` call sites.

File: service/liquidator/src/swap/ref.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/swap/ref.rs]: Fixed. `RefSwap` now stores a `NonceTracker` field and uses it at both `get_access_key_data` call sites (storage registration + swap submission).

File: service/liquidator/src/config.rs [hunk 15]
```diff
@@ -150,6 +155,18 @@ pub struct Args {
...
+
+    /// Telegram thread/topic ID for sending to specific threads in supergroups
+    #[arg(long, env = "TELEGRAM_THREAD_ID")]
+    pub telegram_thread_id: Option<i64>,
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/config.rs]: Fixed. Introduced `SecretString` newtype that prints `<redacted>` in Debug output. `TelegramConfig.bot_token` is now `SecretString`, so any `{:?}` formatting or panic will not expose the token. Added tests verifying redaction.

File: service/liquidator/src/format.rs [hunk 16]
```diff
@@ -4,6 +4,174 @@
...
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/format.rs]: Fixed. Changed `&inner[..17]` to `inner.chars().take(17).collect()` for safe truncation on multi-byte UTF-8.

File: service/liquidator/src/format.rs [hunk 17]
```diff
@@ -159,4 +327,78 @@ mod tests {
...
+    fn test_format_profit_short() {
+        assert_eq!(
+            format_profit_short(952_425, 12_000_000, 6, "nep141:btc.omft.near"),
+            "+0.952425 BTC (+7.9%)"
+        );
+        assert_eq!(
+            format_profit_short(-500_000, 12_000_000, 6, "nep141:wrap.near"),
+            "-0.500000 wNEAR (-4.2%)"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/format.rs:404]: Fixed. Added tests for recursive `nep245:contract:nep141:btc.omft.near` → `BTC` and `format_profit_short` with zero `liquidation_amount` (division-by-zero guard).

File: service/liquidator/src/config.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/config.rs]: Fixed. Changed `telegram_thread_id` from `Option<i64>` with `value_parser` to a plain `String` with `default_value = ""`. Parsing happens in `build_config()` via `parse_optional_i64()` which maps empty/whitespace to `None` and panics with a clear message on invalid input. This avoids the clap parsing issue with empty env vars entirely. Added tests for empty, valid, and invalid inputs.

File: service/liquidator/src/notifier.rs [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+    /// Spawns the send on a background task so callers never block.
+    fn spawn_send(self: &Arc<Self>, message: String) {
+        if self.telegram.is_none() {
+            return;
+        }
+        let this = Arc::clone(self);
+        tokio::spawn(async move {
+            this.send(&message).await;
+        });
+    }
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs:258]: Fixed. Added a bounded `Arc<Semaphore>` (10 permits) to `Notifier`. `spawn_send()` now uses `try_acquire_owned()` — if all permits are taken, the notification is dropped with a warning log instead of spawning unbounded tasks. Permits are released when the background send completes.

File: service/liquidator/src/rpc.rs [hunk 19]
```diff
@@ -135,18 +178,24 @@ pub async fn get_contract_version(
...
 /// # Returns
 ///
 /// Tuple of (nonce, block_hash) to use when constructing a transaction
-#[tracing::instrument(skip(client), level = "debug")]
+#[tracing::instrument(skip(client, nonce_tracker), level = "debug")]
 pub async fn get_access_key_data(
     client: &JsonRpcClient,
     signer: &Signer,
+    nonce_tracker: Option<&NonceTracker>,
 ) -> RpcResult<(u64, CryptoHash)> {
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/rpc.rs]: Fixed. Added `signer` to the `skip()` list: `#[tracing::instrument(skip(client, signer, nonce_tracker), level = "debug")]`. The `Signer` contains the private key and should not appear in trace spans.

**Review by royalf00l — APPROVED** [review]
